File: lab4/laba2/ArrList.cpp
#include "ArrList.h"

int ArrList::constructorCount = 0;
int ArrList::destructorCount = 0;
bool ArrList::isCloning = false;

ArrList::~ArrList()
{
    for (int i = 0; i < capacity; ++i) {
        delete buf[i]; // Освобождаем память для каждого элемента массива
    }
    delete[] buf;
    destructorCount++;
}

ArrList::ArrList() 
{
    if (!isCloning){
        constructorCount++;
    }
    capacity = 5;
    buf = new int*[capacity]; // Выделяем память под массив указателей
    for (int i = 0; i < capacity; ++i) {
        buf[i] = new int; // Выделяем память для каждого элемента массива
    }
    count = 0;
}

int ArrList::getConstructorCount() {
    return constructorCount;
}

int ArrList::getDestructorCount() {
    return destructorCount;
}

void ArrList::Add(int a) {
    if (count >= capacity) {
        // Увеличиваем емкость массива указателей
        int newCapacity = capacity * 2;
        int** newBuf = new int*[newCapacity];
        for (int i = 0; i < newCapacity; ++i) {
            if (i < capacity) {
                newBuf[i] = buf[i]; // Копируем указатели из старого буфера
            } else {
                newBuf[i] = new int; // Выделяем новую память для дополнительных элементов
            }
        }
        delete[] buf; // Освобождаем память для старого массива указателей
        buf = newBuf;
        capacity = newCapacity;
    }
    *(buf[count]) = a; // Записываем значение по адресу, хранящемуся в buf[count]
    count++;
}

void ArrList::Insert(int a, int pos) {
    if (pos < 0 || pos > count) return; 
    if (count >= capacity) {
        // Увеличиваем емкость массива указателей аналогично методу Add
        int newCapacity = capacity * 2;
        int** newBuf = new int*[newCapacity];
        for (int i = 0; i < newCapacity; ++i) {
            if (i < capacity) {
                newBuf[i] = buf[i];
            } else {
                newBuf[i] = new int;
            }
        }
        delete[] buf;
        buf = newBuf;
        capacity = newCapacity;
    }
    for (int i = count; i > pos; --i) {
        *(buf[i]) = *(buf[i - 1]); // Сдвигаем элементы вправо
    }
    *(buf[pos]) = a; // Вставляем новое значение
    count++;
}

void ArrList::Delete(int pos) {
    if (pos < 0 || pos >= count) return;
    for (int i = pos; i < count - 1; ++i) {
        *(buf[i]) = *(buf[i + 1]); // Сдвигаем элементы влево
    }
    count--;
}

void ArrList::Clear() {
    count = 0; // Просто обнуляем количество элементов
}

int& ArrList::operator[](int i) {
    if (i < 0 || i >= count) {
        throw std::out_of_range("Index out of range");
    }
    return *(buf[i]); // Возвращаем значение по адресу, хранящемуся в buf[i]
}

BaseList* ArrList::Clone() {
    ArrList::isCloning = true; // Устанавливаем флаг клонирования
    ArrList* copy = new ArrList();
    copy->Assign(this);
    ArrList::isCloning = false; // Сбрасываем флаг клонирования
    return copy;
}

File: 2labs/ArrList.cs
 public class ArrList : BaseList
{
    private int[] buf; // массив для хранения элементов 
    // private int count;
    // Переопределение свойства для получения количества элементов в списке
    

    public ArrList()
    {
        buf = new int[5]; // задаем нач. размер
        count = 0;
    }

    // метод для добавления элементов в список
    public override void Add(int a)
    {
        if (buf.Length <= count) //проверка заполности буфера
        {
            int[] newBuf = new int[buf.Length * 2]; // создаем новый буфер увеличив его(если надо) и копируем в него элементы
            for (int i = 0; i < buf.Length; i++)
            {
                newBuf[i] = buf[i];
            }
            buf = newBuf;
        }
        //добавляем элемент и увеливаем кол-во элементов
        buf[count] = a;
        count++;
    }

    // метод для добавления элемента на определенную позицию
    public override void Insert(int a, int pos)
    {
        if (pos < 0 || pos > count)
        {
            return; // позиция вне диапазона списка
        }
        if (buf.Length <= count)  //проверка заполности буфера
        {
            // создаем новый буфер увеличив его(если надо) и копируем в него элементы
            int[] newBuf = new int[buf.Length * 2];
            for (int i = 0; i < buf.Length; i++)
            {
                newBuf[i] = buf[i];
            }
            buf = newBuf;
        }
        //сдвигаем элементы вправо, чтобы освободить место
        for (int i = count; i > pos; i--)
        {
            buf[i] = buf[i - 1];
        }
        //вставляем элемент на указанную позицию и увеличиваем кол-во элементов 
        buf[pos] = a;
        count++;
    }

    // метод для удаления элемента с определенной позиции
    public override void Delete(int pos)
    {
        if (pos < 0 || pos >= count)
        {
            return; //проверка позиции
        }
        //сдвигаем элементы влево для удаления
        for (int i = pos; i < count - 1; i++)
        {
            buf[i] = buf[i + 1];
        }
        //уменьшаем кол-во элементов
        count--;
    }

    // метод очистки списка
    public override void Clear()
    {
        for (int i = 0; i < count; i++)
        {
            buf[i] = 0; //зануляем все элементы
        }
        count = 0; //сбрасываем кол-во элементов
    }

    //индексатор для доступа к элементам списка
    public override int this[int i]
    {
        get
        {
            if (i >= count || i < 0)
            {
                throw new ArgumentOutOfRangeException("Element is out of range");
            }
            return buf[i]; //возращаем элемент списка по указанному индексу
        }

        set
        {
            if (i >= count || i < 0)
            {
                //throw new ArgumentOutOfRangeException("Element is out of range");
                return;
            }
            buf[i] = value; //присваиваем новое значение элементу списка по указанному индексу
        }
    }
    // переопределение метода EmptyClone для создания пустой копии текущего списка
    protected override BaseList EmptyClone()
    {
        // создание нового объекта ArrList в качестве пустой копии списка.
        ArrList copy = new ArrList();
        return copy;
    }
}